h2o-core/src/main/java/water/api/ParseHandler.java

```diff
@@ -25,10 +29,18 @@ public ParseV3 parse(int version, ParseV3 parse) {
                                       parse.decrypt_tool != null ? parse.decrypt_tool.key() : null, parse.skipped_columns,
                                       parse.custom_non_data_line_markers != null ? parse.custom_non_data_line_markers.getBytes(): null);
 
-    if (parse.source_frames == null) throw new H2OIllegalArgumentException("Data for Frame '" + parse.destination_frame.name + "' is not available. Please check that the path is valid (for all H2O nodes).'");
+    if (parse.source_frames == null)
+      throw new H2OIllegalArgumentException("Data for Frame '" + parse.destination_frame.name + "' is not available. Please check that the path is valid (for all H2O nodes).'");
     Key[] srcs = new Key[parse.source_frames.length];
-    for (int i = 0; i < parse.source_frames.length; i++)
+    for (int i = 0; i < parse.source_frames.length; i++) {
       srcs[i] = parse.source_frames[i].key();
+    }
+
+    if (parse.partition_by != null) {
+      final String[][] partitionValues = syntheticColumValuesFromPartitions(parse.partition_by, srcs);
```

Comment:
Avoiding the constructor hell here and only adding the `partition_by` to parseSetup, as this is necessary for client exposure. The rest can be done at one place like this.